HPCC-32982 Add alternative span scopes & OwnedSpanScope refactoring

jpmcmu opened the pull request:
- Added ActiveSpanScope to track only current active span
- Renamed existing OwnedSpanScope to OwnedActiveSpanScope
- Added OwnedSpanScope to control only span lifetime

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [X] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [X] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jpmcmu commented on system/jlib/jtrace.hpp:
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
     virtual const char* queryLocalId() const = 0;
 };
 
+//------------------------------------------------------------------------------
```
I felt ActiveSpanScope and OwnedActiveSpanScope better explained what these classes are doing but I am certainly open to changing the names.

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
+    ISpan * prevSpan = nullptr;
+};
+
 class jlib_decl OwnedSpanScope
```
Suggest reordering this before OwnedActiveSpanScope, so there is a logical progression of functionality within the classes.

ghalliday replied:
I almost wonder if OwnedSpanLifetime would be a better name - since it isn't really associated with a scope.

ghalliday commented on system/jlib/jtrace.cpp:
```diff
@@ -1504,13 +1504,36 @@ ISpan * CTraceManager::createServerSpan(const char * name, const IProperties * h
+    ISpan* current = queryThreadedActiveSpan();
+    if (current != span)
+    {
+        const char* currSpanID = current != nullptr ? current->querySpanId() : "null";
```
minor: queryThreadedActiveSpan() can never be null, but CNullSpan::querySpanId() should be changed to return "null".

ghalliday commented on system/jlib/jtrace.hpp:
```diff
@@ -159,11 +159,70 @@ interface ISpan : extends IInterface
+// OwnedActiveSpanScope controls the lifetime of its ISpan while ActiveSpanScope
+// does not. In cases where the ISpan will be used from a single thread and within
+// a single scope OwnedActiveSpanScope should be used. For more complicated scenarios,
+// involving multiple threads, time sliced work, etc ActiveSpanScope should be used.
```
Add an extra comment?  (Do you agree with it?)

involving multiple threads,  time sliced work, ...

a OwnedSpanScope should be stored as a member in a class, and an ActiveSpanScope used to associate that span with each processing thread.

ghalliday reviewed (commented):
@jpmcmu looks good.  A few minor comments.

ghalliday reviewed (approved):
@jpmcmu that looks good.  This will be a really helpful distinction.  I think there is one place in esp that should be revisited now these classes exist - I'll create a jira to cover that.
Please squash and I will merge.

jpmcmu committed:
HPCC-32982 JTrace Add alterntative span socpes & OwnedSpanScope refactoring

- Added ActiveSpanScope to track only current active span
- Renamed existing OwnedSpanScope to OwnedActiveSpanScope
- Added OwnedSpanScope to control only span lifetime

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

jpmcmu commented:
@ghalliday squashed
